### PR TITLE
GH-43455: [Go] Add Opaque canonical extension type

### LINKED ISSUE #43455: [Go] Add Opaque canonical extension type
### Describe the enhancement requested

Part of https://github.com/apache/arrow/issues/43453

### Component(s)

Go

### PR DESCRIPTION
**Pull request by lidavidm**
### Rationale for this change

Add the newly ratified extension type.

### What changes are included in this PR?

The Go implementation only.

### Are these changes tested?

Yes

### Are there any user-facing changes?

No.
* GitHub Issue: #43455

### CODE REVIEW

**Review by zeroshade — APPROVED** [review]

**Comment by lidavidm** [comment]
rebased + fixed the author before I merge

File: go/arrow/extensions/opaque.go [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+// OpaqueType is a placeholder for a type from an external (usually
+// non-Arrow) system that could not be interpreted.
+type OpaqueType struct {
+	arrow.ExtensionBase `json:"-"`
+
+	TypeName   string `json:"type_name"`
+	VendorName string `json:"vendor_name"`
+}
+
+func NewOpaqueType(storageType arrow.DataType, name, vendorName string) *OpaqueType {
```
Comment (joellubi) [inline go/arrow/extensions/opaque.go:38]:
```suggestion
// NewOpaqueType creates a new OpaqueType with the provided storage type, type name, and vendor name.
func NewOpaqueType(storageType arrow.DataType, name, vendorName string) *OpaqueType {
```

**Comment by lidavidm** [comment]
Any further comments here?

**Review by joellubi — APPROVED** [review]
LGTM!

**Pull request merged**